File: app/src/main/java/ru/michanic/mymot/UI/Activities/FilterRegionsActivity.java
package ru.michanic.mymot.UI.Activities;

import android.app.SearchManager;
import android.content.Context;
import android.os.Bundle;

import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.widget.ExpandableListView;
import android.widget.SearchView;

import java.util.List;

import ru.michanic.mymot.Interactors.ApiInteractor;
import ru.michanic.mymot.Models.Location;
import ru.michanic.mymot.MyMotApplication;
import ru.michanic.mymot.Protocols.LoadingInterface;
import ru.michanic.mymot.R;
import ru.michanic.mymot.UI.Adapters.RegionsExpandableListAdapter;
import ru.michanic.mymot.Utils.DataManager;

public class FilterRegionsActivity extends UniversalActivity {

    private RegionsExpandableListAdapter regionsExpandableListAdapter;
    private ApiInteractor apiInteractor = new ApiInteractor();

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_filter_regions);
        setNavigationTitle("Регион");

        final List<Location> regions = new DataManager().getRegions();
        regionsExpandableListAdapter = new RegionsExpandableListAdapter(regions, this);

        final ExpandableListView expandableListView = (ExpandableListView) findViewById(R.id.expandable_list_view);
        expandableListView.setAdapter(regionsExpandableListAdapter);

        expandableListView.setOnGroupClickListener(new ExpandableListView.OnGroupClickListener() {
            @Override
            public boolean onGroupClick(ExpandableListView parent, View v, final int groupPosition, long id) {
                if (groupPosition == 0) {
                    MyMotApplication.searchManager.setRegion(null);
                } else {
                    Location region = regions.get(groupPosition - 1);
                    //Log.e("getCitiesCount", String.valueOf(region.getCitiesCount()));
                    //regionsExpandableListAdapter.notifyDataSetChanged();
                    if (region.getCitiesCount() == 0) {
                        apiInteractor.loadRegionCities(region, new LoadingInterface() {
                            @Override
                            public void onLoaded() {
                                Log.e("expandGroup", String.valueOf(groupPosition));
                                expandableListView.collapseGroup(groupPosition);
                                expandableListView.expandGroup(groupPosition);
                            }
                            @Override
                            public void onFailed() {

                            }
                        });
                    }
                }
                return false;
            }
        });

        expandableListView.setOnChildClickListener(new ExpandableListView.OnChildClickListener() {
            @Override
            public boolean onChildClick(ExpandableListView parent, View v, int groupPosition, int childPosition, long id) {
                if (childPosition == 0) {
                    Location region = regions.get(groupPosition - 1);
                    MyMotApplication.searchManager.setRegion(region);
                } else {
                    Location region = regions.get(groupPosition - 1);
                    Location city = region.getCities().get(childPosition - 1);
                    MyMotApplication.searchManager.setRegion(city);
                }
                return false;
            }
        });
    }

}

File: app/src/main/java/ru/michanic/mymot/Kotlin/UI/Activities/FilterRegionsActivity.kt
package ru.michanic.mymot.Kotlin.UI.Activities

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ExpandableListView
import ru.michanic.mymot.Kotlin.Interactors.ApiInteractor
import ru.michanic.mymot.Kotlin.MyMotApplication
import ru.michanic.mymot.Kotlin.Protocols.LoadingInterface
import ru.michanic.mymot.R
import ru.michanic.mymot.Kotlin.UI.Adapters.RegionsExpandableListAdapter
import ru.michanic.mymot.Kotlin.Utils.DataManager

class FilterRegionsActivity : UniversalActivity() {
    private var regionsExpandableListAdapter: RegionsExpandableListAdapter? = null
    private val apiInteractor = ApiInteractor()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_filter_regions)
        setNavigationTitle("Регион")
        var expandGroupPosition = 0
        var expandedRegionId = 0
        val filterRegion = MyMotApplication.searchManager?.region
        if (filterRegion != null) {
            val cityRegion = MyMotApplication.dataManager?.getRegionById(filterRegion.regionId)
            expandedRegionId = cityRegion?.id ?: filterRegion.id
        }
        val regions = DataManager().regions
        if (expandedRegionId > 0) {
            var row = 1
            for (region in regions) {
                val citiesCount = MyMotApplication.dataManager?.getRegionCitiesCount(region.id)
                if (expandedRegionId == region.id) {
                    expandGroupPosition = row
                }
                row++
            }
        }
        regionsExpandableListAdapter = RegionsExpandableListAdapter(regions, this)
        val expandableListView = findViewById<View>(R.id.expandable_list_view) as ExpandableListView
        expandableListView.setAdapter(regionsExpandableListAdapter)
        expandableListView.setOnGroupClickListener { parent, v, groupPosition, id ->
            if (groupPosition == 0) {
                MyMotApplication.searchManager?.region = null
                onBackPressed()
            } else {
                val region = regions[groupPosition - 1]
                val citiesCount = MyMotApplication.dataManager?.getRegionCitiesCount(region.id)
                if (citiesCount == 0) {
                    apiInteractor.loadRegionCities(region, object : LoadingInterface {
                        override fun onLoaded() {
                            Log.e("expandGroup", groupPosition.toString())
                            expandableListView.collapseGroup(groupPosition)
                            expandableListView.expandGroup(groupPosition)
                        }

                        override fun onFailed() {}
                    })
                }
            }
            false
        }
        expandableListView.setOnChildClickListener { parent, v, groupPosition, childPosition, id ->
            if (childPosition == 0) {
                val region = regions[groupPosition - 1]
                MyMotApplication.searchManager?.region = region
            } else {
                val region = regions[groupPosition - 1]
                val city =
                    MyMotApplication.dataManager?.getRegionCities(region.id)?.get(childPosition - 1)
                MyMotApplication.searchManager?.region = city
            }
            onBackPressed()
            false
        }
        if (expandGroupPosition > 0) {
            expandableListView.expandGroup(expandGroupPosition)
            expandableListView.setSelectedGroup(expandGroupPosition)
        }
    }
}